### DISCUSSION TITLE
Workflow error : workflow reference count exceeds the limit of 20

### DISCUSSION THREAD

**Discussion by ghost** [community/community#39394] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

We have a workflow and when executing that we get below error -
 
_"workflow reference count exceeds the limit of 20"_
 
I am not able to find any document which talk about this reference count limit and what does workflow reference means.

**Comment by heoelri** [comment]
We hit the same limit last week. I've raised a support ticket to figure out of this is a soft limit that can be raised or not. Still waiting for a response.

**Comment by ghost** [comment]
@heoelri, please can you share the support ticket here.

**Reply by heoelri** [reply]
My TicketId is #1892877

**Comment by MatanHeledPort** [comment]
Hey!
Any updates about this?
I we just encountered the same issue.
Seems like this is because of to many references to workflows

EDIT:
https://github.com/orgs/community/discussions/32918

**Reply by heoelri** [reply]
The response we got from the GitHub support was:

> While **we don't currently have that functionality implemented**, I've gone ahead and added your organization's voice to the internal issue tracking the feature request which I've raised on your organization's behalf. At this time **we can't say when or if it will be released** but we'd recommend that you keep an eye on the GitHub Changelog where updates are announced.
> 11/30/2022 GitHub Support

**Reply by MatanHeledPort** [reply]
Thanks @heoelri 
Looking forward for updates

Matan

**Reply by sukhdevtech9** [reply]
Any updates on this ? We have run into similar situation in our project. Looking forward to ideas on how to resolve this. Thanks

**Comment by cjlapao** [comment]
I have now encountered the same limitation on this, seems odd the limit as there is not a lot of documentation to avoid it in the first place, it is now difficult to go back and redo the approach when you already invested a lot of time in the separation of concerns

**Comment by vppawar** [comment]
Any updates on this ? We have run into similar situation in our project. Please point to the reference which explains to resolve this issue. TIA!

**Comment by thatkookooguy** [comment]
Same here. just encountered this and can't find a way to fix this. Is there any official response?

**Comment by kevintijssen** [comment]
Unfortunate this limit is a reference to this artical. I'm not sure if there going to increase this.
https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/reusing-workflows#limitations

**Comment by cbugneac-nex** [comment]
We also hit this limitation. In order to not have a lot of duplicated code we have to reference reusable workflows but bitten by this limitation.

**Comment by laurentpellegrino** [comment]
Any chance to have this limit increased? In my case I need about 40 :X

**Comment by christiangda** [comment]
We have the same issue with this limit, so is there a way to increase that?

I am still waiting for a formal answer from Github.com. 
![Screenshot 2024-07-03 at 10 07 05](https://github.com/community/community/assets/1197820/0b723cbb-1d4f-4bab-b709-367de6ff0bb3)


https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/reusing-workflows#limitations

**Comment by Paebbels** [comment]
I also run into this limit.

As mentioned in https://github.com/orgs/community/discussions/39394#discussioncomment-6371560, GitHub advocates to apply *reusable workflows* for better structuring and reduction of duplicate code. A limit of 20 doesn't seem to be a reasonable limit. Just think of a Bash/Python script limited to 20 functions...

A typical Python pipeline needs:
1. pipeline preparation
1. type checks
1. linting
1. unit tests
1. code coverage
1. documentation
1. translate to PDF using latex
1. publish test results
1. publish to github pages
1. publish to PyPI
1. package wheel
1. install and check package after installation
1. create release page

This basic simple pipeline uses already 13 workflow templates.

The number ob templates/jobs increates due to missing features and bugs in GitHub Actions:
* A preparation step is needed because GHA doesn't support variables properly like e.g. GitLab.  
* More over certain environment variables are not provided as individual information, thus a job and script is needed to split information (organisation vs. repository name, reference vs. branch/tag name, ...).
* It's not possible to defined *needs*-rules to job-matrix members, thus a workflow needs to be instantiated multiple times for e.g. *fast* and *slow* job-matrix members. This allows e.g. a documentation job to continue while other matrix  items are still executing or even waiting. Especially, macOS jobs are very very slow (waiting times + execution times).

To summarize: Besides a low limit of 20 reusable workflows, the problem gets increased by missing features in GitHub Actions itself, were a workaround needs to add more workflow templates.

If needed, a limit of 100 or 50 should be applied or an option should be provided for bigger open-source projects to request a limit-change.

Example pipelines of mine:
* GHDL (Ada/GCC): **120 jobs** in 30 minutes - https://github.com/Paebbels/ghdl/actions/runs/15261120508
* pyTooling (Python): **54 jobs** in 7 minutes - https://github.com/pyTooling/pyTooling/actions/runs/15220158114
